util/hot_backup.cc

```diff
@@ -0,0 +1,654 @@
+
+    // 1 means CheckHotBackupTrigger()'s counter is only
+    //   one left
+    if (1==ret_val)
+    {
+        // hmm, call file sync() at this point?
+        Log(NULL, "leveldb HotBackup complete.");
+        ResetTrigger();
+    }   // if
+
+    return;
+
+}   // HotBackupFinished
+
+
+/**
+ * Routine called by DBList()->ScanDBs.  This has full context of one open
+ *  leveldb database.  Validates then schedules a backup job via compaction
+ *  threads.
+ */
+void
+DBImpl::HotBackup()
+{
+    bool create_backup_event(false);
+
+    // hold mutex while reviewing database state
+    //  (note:  never good to call Log() while holding mutex
+    //   but these are unlikely error cases)
+    {
+        // get mutex,
+        MutexLock l(&mutex_);
+
+        // set backup flag && see if
+        assert(false==hotbackup_pending_);
+        if (!hotbackup_pending_)
+        {
+            // test if shutting_down,
+            if (!shutting_down_.Acquire_Load())
+            {
+                create_backup_event=true;
+                hotbackup_pending_=true;
+            }   // if
+            else
+            {
+                Log(options_.info_log, "HotBackup aborted, shutdown in progress");
+            }   // else
+        }   // if
+        else
+        {
+            Log(options_.info_log, "HotBackup aborted, backup already in progress");
+        }   // else
+    }   // mutex released
+
+    // post event and log message outside mutex
+    if (create_backup_event)
+    {
+        // increment pending backup count
+        gHotBackup->HotBackupScheduled();
+
+        // schedule backup job within compaction queue
+        ThreadTask * task=new HotBackupTask(this);
+        gCompactionThreads->Submit(task, true);
+        Log(options_.info_log, "HotBackup queued");
+    }   // if
+
+    return;
+
+}   // DBImpl::HotBackup
+
+
+void
+HotBackupTask::operator()()
+{
+    bool good;
+    long log_position(0);
+
+    // rotate directories (tiered storage reminder)
+    good=gHotBackup->PrepareDirectories(m_DBImpl.GetOptions());
+
+    // purge imm or current write buffer
+    good=good && m_DBImpl.PurgeWriteBuffer();
+
+    // Gather random data items for MANIFEST and
+    //  other supporting files.
+    if (good)
+    {
+        // get LOG file size
+        if (NULL!=m_DBImpl.GetLogger())
+            log_position=m_DBImpl.GetLogger()->LogSize();
+        else
+            log_position=0;
+
+        // retrieve current version and increment refs_ (requires mutex)
+        //  (includes code that creates hard links to .sst files)
+        good=m_DBImpl.WriteBackupManifest();
+
+        // last step is to replicate as much of LOG file from time of
+        //  backup (close, but not exact)
+        //  Assumes that 0 for log_position implies user has a custom
+        //  Logger class that this routine will fail to copy.
+        if (good && 0!=log_position)
+            good=m_DBImpl.CopyLOGSegment(log_position);
```

Comment:
LOG file is best efforts basis.  It is not essential for reuse of the data.  In fact, it gets moved to LOG.old and a new LOG file installed upon start of leveldb.  The underlying question is how do we know when ftell() gets updated:  before a printf to LOG, after, or some unknown time.  I am not going to try and solve that.